torch/_tensor.py

```diff
@@ -904,6 +904,55 @@ def rename(self, *names, **rename_map):
         # See Note [rename_ / rename API]
         return update_names(self, names, rename_map, inplace=False)
 
+    def to_sparse_csr(self):
+        """ Convert a tensor to compressed row storage format. Only works with 2D tensors.
+
+        Examples::
+
+            >>> dense = torch.randn(5, 5)
+            >>> sparse = dense.to_sparse_csr()
+            >>> sparse._nnz()
+            25
+
+        """
+        shape = self.size()
+        fill_value = 0
+        if len(shape) != 2:
+            raise RuntimeError("Only 2D tensors can be converted to the CSR format but got shape: ", shape)
+
+        if self.is_sparse:
+            coalesced_self = self.coalesce()
+            row_indices = coalesced_self.indices()[0]
+            ro = [0]
+            i = 0
+            for irow in range(self.shape[0]):
+                while i < row_indices.size()[0] and row_indices[i] == irow:
+                    i += 1
+                ro.append(i)
+
+            return torch.sparse_csr_tensor(torch.tensor(ro, dtype=row_indices.dtype),
+                                           coalesced_self.indices()[1], coalesced_self.values(),
+                                           size=coalesced_self.shape, dtype=coalesced_self.dtype)
+        elif self.is_sparse_csr:
+            return self
+        else:
+            ro = [0]
+            co = []
+            vals: List[Any] = []
+
+            for irow in range(shape[0]):
+                row = self[irow, :]
+                selection = row != fill_value
+                select_nums = row[selection]
+                ro.append(ro[-1] + len(select_nums))
+                co.extend(torch.nonzero(selection).flatten().tolist())
+                vals.extend(select_nums.flatten().tolist())
+
+            return torch.sparse_csr_tensor(torch.tensor(ro, dtype=torch.int32), 
+                                           torch.tensor(co, dtype=torch.int32),
```

Comment:
Awareness note: the default dtype for indices should depend on the used MKL variant: ILP64 or LP64.

I am not sure if there is a direct way to determine from Python which variant of MKL is used. So I suggest adding the corresponding comment that the indices of the CSR tensor have suboptimal dtype when ILP64 is used.

Reply:
Cannot determine from python directly. It will need an extra native function, which I don't think is a good idea to satisfy just this use case since we internally cast it to 64 bit int if the user calls MKL functions anyway. I've removed this code now so it won't make a difference.